src/util/ConfigManager/ConfigOptionProxy.h

```diff
@@ -0,0 +1,96 @@
+// Copyright 2023, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Andre Schlegel (March of 2023, schlegea@informatik.uni-freiburg.de)
+
+#pragma once
+
+#include <absl/strings/str_cat.h>
+
+#include <concepts>
+#include <stdexcept>
+#include <type_traits>
+
+#include "util/ConfigManager/ConfigOption.h"
+#include "util/Exception.h"
+
+namespace ad_utility {
+namespace detail {
+/*
+@brief Implementation of proxy/reference to a `ConfigOption`. Also saves the
+type of value, that the configuration option holds. Needed, because without a
+`ConfigOptionProxy` and `ConstConfigOptionProxy`, people could easily discard
+the `const` from a `const ConfigOptionProxy` by copying it, to have non const
+access to the referenced config option.
+
+@tparam T The type of value, the later given config option will hold.
+@tparam ConfigOptionType The kind of config option, this proxy will reference
+to. Must be `ConfigOption`, or `const ConfigOption`.
+*/
+template <typename T, typename ConfigOptionType>
+requires ad_utility::isTypeContainedIn<T, ConfigOption::AvailableTypes> &&
+         (std::same_as<std::decay_t<ConfigOptionType>, ConfigOption> ||
+          std::same_as<std::decay_t<ConfigOptionType>, const ConfigOption>)
+class ConfigOptionProxyImplementation {
+  ConfigOptionType* option_ = nullptr;
+
+ public:
+  using Type = T;
+
+  // Custom constructor.
+  explicit ConfigOptionProxyImplementation(ConfigOptionType& opt)
+      : option_(&opt) {
+    // Make sure, that the given `ConfigOption` holds values of the right type.
+    if (!opt.template holdsType<T>()) {
+      throw std::runtime_error(absl::StrCat(
+          "Error while creating 'ConfigOptionProxy' for 'ConfigOption' '",
+          opt.getIdentifier(),
+          "': Mismatch between the template type of proxy and the type of "
+          "values, the configuration option can hold."));
+    }
+  }
+
+  // Get access to the configuration option, this is a proxy for. Const access
+  // only for the const version, to make the usage clearer.
+  ConfigOptionType& getConfigOption() const
```

Comment:
If this always returns `ConfigOption&` then you don't need the
`requires` clause, but then a non-const proxy with an external `const` qualifier works as expected.

Reply:
This function is only for `NonConstConfigOptionProxy`, which should **never** be `const`, because that would muddle the wanted usage.

If people want a `const` proxy, they have to use `ConstConfigOptionProxy`. Which is also why `NonConstConfigOptionProxy` isn't really usable, when made `const`.

Reply:
Alright, this is not that important for me.

Reply:
Alright then.